### ISSUE TITLE
dlib::cublas error: CUDA Runtime API initialization failed using cnn to extract face descriptors

### ISSUE THREAD

**Issue by ndtreviv** [davisking/dlib#2200]
## Expected Behavior

I'm attempting to use the cnn face detector to extract face descriptors:

ie:
```
void extract_faces_cnn(std::vector<dlib::matrix<rgb_pixel>> &faces, std::vector<dlib::rectangle> &coords, dlib::cv_image<rgb_pixel> &img, shape_predictor &sp) {
        matrix<rgb_pixel> imgmat = mat(img);

    while(imgmat.size() < 1800*1800)
    {
        pyramid_up(imgmat);
    }

    auto dets = cnn_detector(imgmat);
    for (auto&& face : dets)
    {
        auto shape = sp(imgmat, face); 
        matrix<rgb_pixel> face_chip;
        extract_image_chip(img, get_face_chip_details(shape,150,0.25), face_chip);
        faces.push_back(move(face_chip));
        coords.push_back(face.rect);
    }
}
```

I would expect the face descriptors to be extracted using the cnn detector, which in turn utilises cuDNN running on the GPU in order to speed things up.

## Current Behavior

I get this error:

```
terminate called after throwing an instance of 'dlib::cublas_error'
  what():  Error while calling cublasCreate(&handles[new_device_id]) in file /home/ubuntu/dlib/dlib/cuda/cublas_dlibapi.cpp:78. code: 1, reason: CUDA Runtime API initialization failed.
```

## Steps to Reproduce

on Ubuntu 20.04:

1. apt-get install gfortran libopenblas-dev liblapack-dev nvidia-cuda-toolkit
2. Install CUDA from repo: http://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/
3. Test nvcc --version:
```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243
```
4. Download and install cuDNN from nvidia, installing:
```
$ dpkg -i libcudnn8_8.0.4.30-1+cuda10.1_amd64.deb
$ dpkg -i libcudnn8-dev_8.0.4.30-1+cuda10.1_amd64.deb
```
5. Check `nvidia-smi`:
```
| NVIDIA-SMI 455.23.05    Driver Version: 455.23.05    CUDA Version: 11.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:1E.0 Off |                    0 |
| N/A   60C    P0    27W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```
6. Clone, build and install dlib:
```
git clone https://github.com/davisking/dlib.git && \
cd dlib && \
mkdir build && cd build  && \
cmake -DUSE_AVX_INSTRUCTIONS=ON -DBUILD_SHARED_LIBS=1 -DCUDA_HOST_COMPILER=/usr/bin/gcc-8 ../ && \
cmake --build . --config Release && \
make install
```
Output of cmake is:
```
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using CMake version: 3.16.3
-- Compiling dlib version: 19.21.99
-- Enabling AVX instructions
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found X11: /usr/include   
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Found system copy of libpng: /usr/lib/x86_64-linux-gnu/libpng.so;/usr/lib/x86_64-linux-gnu/libz.so
-- Searching for BLAS and LAPACK
-- Searching for BLAS and LAPACK
-- Found PkgConfig: /usr/bin/pkg-config (found version \"0.29.1\") 
-- Checking for module 'cblas'
--   No package 'cblas' found
-- Checking for module 'lapack'
--   Found lapack, version 0.3.8+ds
-- Looking for cblas_ddot
-- Looking for cblas_ddot - not found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Found OpenBLAS library
-- Looking for sgetrf_single
-- Looking for sgetrf_single - found
-- Using OpenBLAS's built in LAPACK
-- BLAS library does not have cblas symbols, so dlib will not use BLAS or LAPACK
-- Found BLAS: /usr/lib/x86_64-linux-gnu/libopenblas.so  
-- Looking for sgesv
-- Looking for sgesv - not found
-- Looking for sgesv_
-- Looking for sgesv_ - found
-- Found CUDA: /usr (found suitable version \"10.1\", minimum required is \"7.5\") 
-- Looking for cuDNN install...
-- Found cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so
-- Building a CUDA test project to see if your compiler is compatible with CUDA...
-- Building a cuDNN test project to check if you have the right version of cuDNN installed...
-- Enabling CUDA support for dlib.  DLIB WILL USE CUDA
-- C++11 activated.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ubuntu/dlib/build
```
7. Run the cuDNN tests, just to make sure:
```
~/dlib/dlib/cmake_utils/test_for_cudnn/build$ cmake -DUSE_AVX_INSTRUCTIONS=ON ../
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr (found suitable version "10.1", minimum required is "7.5") 
-- Looking for cuDNN install...
-- Found cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so
-- C++11 activated.
-- Configuring done
-- Generating done

~/dlib/dlib/cmake_utils/test_for_cudnn/build$ make
Scanning dependencies of target cudnn_test
[ 50%] Building CXX object CMakeFiles/cudnn_test.dir/home/ubuntu/dlib/dlib/cuda/cudnn_dlibapi.cpp.o
[100%] Linking CXX static library libcudnn_test.a
[100%] Built target cudnn_test
```
(They pass, which makes me believe that everything should work ok.)

8. Build and install my program using CMake, linking in dlib.
CMakeLists.txt:
```
cmake_minimum_required(VERSION 2.8.12)

project(dlib-jni)

set(CMAKE_CXX_STANDARD 11)
set(USE_AVX_INSTRUCTIONS ON CACHE BOOL "Use AVX instructions")
#set(CMAKE_TOOLCHAIN_FILE /home/ubuntu/vcpkg/scripts/buildsystems/vcpkg.cmake CACHE STRING "Vcpkg toolchain file")

add_definitions(-DOPENCV_TRAITS_ENABLE_DEPRECATED)
find_package(OpenCV 3 REQUIRED)
include_directories(${OpenCV_INCLUDE_DIRS})

find_package(dlib REQUIRED) #possibly supporting components!
message(STATUS "Using dlib-${dlib_VERSION}")

set(BLA_VENDOR OpenBLAS)
find_package(BLAS REQUIRED)
include_directories(${BLAS_INCLUDE_DIRS})

include_directories(src)

set(DLIB_JNI_THIRDPARTY_LIBS opencv_highgui
                        opencv_imgcodecs
                        opencv_imgproc
                        opencv_core
                        lapack)

find_package(JNI REQUIRED)
message (STATUS "JNI_INCLUDE_DIRS=${JNI_INCLUDE_DIRS}")
message (STATUS "JAVA_INCLUDE_PATH =${JAVA_INCLUDE_PATH}")
message (STATUS "JNI_LIBRARIES=${JNI_LIBRARIES}")
message (STATUS "JAVA_JVM_LIBRARY=${JAVA_JVM_LIBRARY}")
include_directories(${JNI_INCLUDE_DIRS})

add_library(dlib_jni SHARED src/dlib-jni.cpp)
set_target_properties(dlib_jni PROPERTIES POSITION_INDEPENDENT_CODE ON) # This made it work
target_link_libraries(dlib_jni dlib::dlib ${DLIB_JNI_THIRDPARTY_LIBS} ${JAVA_JVM_LIBRARY})
```
Command to build:
```
$ cd dlib-jni && mkdir build && cd build && cmake -DUSE_AVX_INSTRUCTIONS=ON ../ && \
cmake --build . --config Release && \
sudo cp *.so /usr/local/lib/
```
9. Run the program.

* **Version**: 19.21.99
* **Where did you get dlib**: Cloned from master on github repo 2nd October
* **Platform**: 
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
```
* **Compiler**: CUDA_HOST_COMPILER is set to use gcc-8 as per my cmake command above. The compiler used according to cmake output is:
```
gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by ndtreviv** [comment]
In addition to this, I installed gcc-8 and g++-8 on the box, and changed the cmake commands to include: `-DCMAKE_C_COMPILER=/usr/bin/gcc-8 -DCMAKE_CXX_COMPILER=/usr/bin/g++-8`

I still get the same error.

**Comment by davisking** [comment]
This isn't because of anything in dlib.  You have to follow the nvidia instructions, especially regarding what platforms are supported by which versions of cuda.  see https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html.

**Comment by davisking** [comment]
Also, some of the stuff you posted says it's running cuda 10.1 and others 11.1.  You probably have a messed up cuda install.

**Comment by ndtreviv** [comment]
Thanks for your help. FWIW nothing with CUDA install was messed up: https://stackoverflow.com/questions/53422407/different-cuda-versions-shown-by-nvcc-and-nvidia-smi

I'll go back over the instructions from nvidia. We're running other stuff on the GPU here, so I'm fairly sure that the environment is OK, but clearly something is wrong. I'll report back for the purposes of educating other people coming here from Google.

**Comment by ndtreviv** [comment]
When I install cuDNN 11, I get this:

```
Ahhh…shame
terminate called after throwing an instance of 'dlib::cudnn_error'
  what():  Error while calling cudnnFindConvolutionBackwardFilterAlgorithm( context(), descriptor(data), descriptor(dest_desc), (const cudnnConvolutionDescriptor_t)conv_handle, (const cudnnFilterDescriptor_t)filter_handle, num_possible_algorithms, &num_algorithms, perf_results.data()) in file /home/ubuntu/dlib/dlib/cuda/cudnn_dlibapi.cpp:885. code: 2, reason: CUDA Resources could not be allocated.
```

**Comment by ndtreviv** [comment]
I followed all the CUDA install instructions (https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#ubuntu-installation), and the post-install instructions (https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#post-installation-actions).

Eventually, the issue was that I needed to restart (possibly after setting the LD_LIBRARY_PATH).

**Comment by ndtreviv** [comment]
For people coming here from Google, I've finally bottomed all of this out:

# TL;DR:
1. Make sure you run the post-install instructions
2. Don't install nvidia-cuda-toolkit
3. On Ubuntu you don't need to reboot after installing CUDA, cuDNN or dlib
4. `CUDA Resources could not be allocated` basically means that you don't have enough GPU left.

# Details
It turns out that I did, in fact, have CUDA 11.1 installed. When I `apt-get install nvidia-cuda-toolkit` it installs nvcc, which then reports CUDA as being CUDA 10.1 - why this is, I still don't know. I installed `nvidia-cuda-toolkit` because I thought the nvcc utility hadn't been installed. This was an incorrect assumption - it was actually that nvcc wasn't on the `PATH` because I hadn't followed the post-installation instructions correctly.

Once I had followed the post-install instructions (linked in previous comment), which involved adding cuda to the `PATH` the nvcc utility became available and reported CUDA 11.1 correctly.

Once I had populated the `PATH` and `LD_LIBRARY_PATH` correctly, as per the post install instructions, I was then getting an error along the lines of:
```
... /home/ubuntu/dlib/dlib/cuda/cudnn_dlibapi.cpp:885. code: 2, reason: CUDA Resources could not be allocated.
```
This was confusing, because the dlib cuDNN tests were passing.

I assumed that the post install instructions hadn't taken effect, and so rebooted the vm I was working on. Everything then ran perfectly fine. For a while. Then it stopped working. Again, the need to reboot was an incorrect assumption. 

It turned out that we had tf-serving running in a docker container on that vm as well. tf-serving was using up almost all of the GPU, but leaving just enough for the cuDNN tests to be able to pass.
My code worked after a reboot because tf-serving hadn't had time to fire up yet. After a while, though, my code stopped working with the same error.

All of this is valid for Ubuntu 20.04.
